Draft sobre Rust, tipos e Owenerships

bltavares opened the pull request:
Apresentação para iniciar sobre definição de tipos própios e uma leve introdução ao modelo de ownership. 

[Rendered](https://github.com/bltavares/presentations/blob/rust-tipos-e-ownerships/rust-tipos-e-ownership/rust-tipos-e-ownership.org)

bltavares committed:
Draft sobre Rust, tipos e Owenerships

romariorios commented on rust-tipos-e-ownership/rust-tipos-e-ownership.org:
```diff
@@ -0,0 +1,438 @@
+* Uma solução
+
+#+BEGIN_SRC rust
+  fn color(flower: String) -> String {
```
Não poderia ser `fn color(flower: &str) -> String` aqui? Até onde sei, `String`s em Rust decaem para `&str`.

bltavares replied:
Você esta certo. Poderia ser sim.

Eu pensei em não introduzir referencias no primeiro exemplo, escrever um código simples para quem está vindo de outras linguagens.
Como eu só introduzo referencias mais pra frente eu deixei pra explicar o `String` e `&str` no final.

romariorios replied:
Eu acho que usar `String` complica mais que simplifica -- vide as linhas que você teve que usar `.to_string()`, que podem confundir o leitor. Você poderia usar `&str` e deixar pra dar os detalhes sobre ele mais tarde -- o que é justamente o que você está fazendo com as chamadas para `.to_string()`.

bltavares replied:
Eu debati se seria melhor escrever `&str` ou usar o `.to_string` e ainda tenho dúvidas.

Para todas as pessoas que não vem de C/C++, `&tipo` é uma syntax bem alienigena.

Será que eu poderia utilizar um outro exemplo, com outra estrutura de dados? Talvez `bool -> i32`?

bltavares replied:
E se for esse o exemplo?

``` rust
fn color(flower: String) -> String {
    if flower == "1" {
        "red".to_string()
    } else if flower == "2" {
        "blue".to_string()
    } else {
      panic!("So far only Roses or Violets") 
    }
}

fn main() {
    // Roses are 1
    // Violets are 2
    let flower = format!("{}", 1);
    let inspect_color = color(flower);

    println!("The color is: {}", inspect_color);
}
```

bltavares replied:
E se eu utilizar `String::from("literal")` e avisar que vou explicar melhor sobre isso no final?

romariorios replied:
Em 12 de janeiro de 2016 17:12, Bruno Lara Tavares <notifications@github.com

> escreveu:
> 
> In rust-tipos-e-ownership/rust-tipos-e-ownership.org
> https://github.com/bltavares/presentations/pull/1#discussion_r49506806:
> 
> > +Vamos ver um pouco sobre /Enums/ e /Structs/ sendo aplicadas em problemas.
> > +
> > +
> > +\* Problema 1
> > +
> > +Você esta fazendo um programa para falar qual a cor de uma flor.
> > +Mas ele so funciona pra /Rosas/ e /Violetas/.
> > +
> > +/Rosas/ são vermelhas.
> > +/Violetas/ são azuis.
> > +
> > +
> > +\* Uma solução
> > +
> > +#+BEGIN_SRC rust
> > -  fn color(flower: String) -> String {
> 
> E se eu utilizar String::from("literal") e avisar que vou explicar melhor
> sobre isso no final?
> 
> Pra mim, continua a mesma coisa.

Minha opinião quanto a esse tipo de coisa é o seguinte: se houver mais de
uma opção de exemplo que pode alienar o leitor por conter algum conceito
relativamente avançado, escolha a que é mais idiomática. Nesse caso, eu
acho que o mais idiomático seria usar `&str` em vez de `String` no
argumento da função. Minha linguagem principal é o C++, então sou um pouco
suspeito pra falar, mas acho que, pra um iniciante, `&str` é tão estranho
quando converter uma string pra string.

Resumindo: eu acho que `&str` é a melhor opção, mas você é quem sabe.

> —
> Reply to this email directly or view it on GitHub
> https://github.com/bltavares/presentations/pull/1/files#r49506806.

## 

Luiz Romário Santana Rios

bltavares replied:
Obrigado pelo feedback.

Concordo que converter uma string pra string é estranho. Eu odeio ver isso no meu código também haha

Eu também sou suspeito pra falar, então vou deixar esse ponto em aberto por enquanto para tentar receber mais feedbacks, principalmente de quem vem de linguagens que não tem `&str`.

carlosgaldino replied:
Talvez seja o caso de deixar claro que não são dois tipos iguais. Então não é "converter de uma string pra string". A conversão é entre `str` para `String`. Acredito que a confusão pode ser acentuada por esse motivo, se fossem dois tipos iguais qual a razão em haver uma conversão? Se são dois tipos diferentes então sim existe uma razão para tal conversão.

Uma ideia poderia ser de explicar a motivação para se ter dois tipos de strings e aí falar sobre a questão dos literals serem estáticos pois são alocados no initialized data section (`.data` do processo) ou na stack quando usado um lifetime diferente de `'static` e por isso tende a ser natural um tipo `String` pra representar strings que não seguem essas regras pois podem crescer e por isso são heap allocated.

hannelita replied:
<2cents> Acharia bacana explicar a diferença entre eles. Nem todo mundo é familiarizado com &str.

bltavares replied:
Pesoal, obrigado pelo feedback.

Decidi manter o uso de `String` na assinatura de função.
Recebi feedbacks de pessoas que nunca viram a sintax `&Tipo` em outras linguagens, que eles ficaram mais confortáveis de ficar com a dúvida do motivo de "literais" não serem Strings.

Vendo o fluxo da apresentação, eu introduzo essa dúvida, e muitas pessoas perguntaram sobre isso. Aos poucos vou introduzindo os conceitos de uso de memória, pra no final mostrar que "literais" são `&'static str` e não `String`.

Vou tentar incrementar a explicação nos slides, para deixar mais claro onde literais são alocados e mais sobre stack e heap.

Obrigado!

bltavares committed:
Adiciona slides falando que é introdução

Os slides tem um pouco mais de informações com dicas do que procurar
depois e onde procurar esse tipo de conteudo.

bltavares committed:
Adiciona uma nota sobre nao ter GC

bltavares committed:
Utiliza u8 ao inves de i32 para melhor representar os valores de cor

hannelita committed:
Typo fix

bltavares committed:
Merge pull request #2 from hannelita/rust-tipos-e-ownerships

Fixing some typos

bltavares committed:
Fix phrase

bltavares committed:
Introduz a assinatura proposta para a função no slide do problema

Dessa forma, podemos focar na assinatura, e assim que encontrarmos o
`"literal".to_string()` a dúvida vai aparecer.

Assim, é introduzido a assinatura que todos possam concordar, e a
dúvida de por que não posso passar um literal, para podermos responder
isso no final.

bltavares committed:
Atualiza exemplos no Playground

bltavares committed:
Adiciona uma indicação que os links são para o Playground.
